lib/matplotlib/backends/backend_gtk3.py

```diff
@@ -699,7 +717,223 @@ def get_filename_from_user (self):
 
         return filename, self.ext
 
-class DialogLineprops(object):
+
+class NavigationGTK3(NavigationBase):
+    pass
+
+
+class RubberbandGTK3(RubberbandBase):
+    def __init__(self, *args, **kwargs):
+        RubberbandBase.__init__(self, *args, **kwargs)
+        self.ctx = None
+
+    def draw_rubberband(self, x0, y0, x1, y1):
+        # 'adapted from http://aspn.activestate.com/ASPN/Cookbook/Python/
+        # Recipe/189744'
+        self.ctx = self.figure.canvas.get_property("window").cairo_create()
+
+        # todo: instead of redrawing the entire figure, copy the part of
+        # the figure that was covered by the previous rubberband rectangle
+        self.figure.canvas.draw()
+
+        height = self.figure.bbox.height
+        y1 = height - y1
+        y0 = height - y0
+        w = abs(x1 - x0)
+        h = abs(y1 - y0)
+        rect = [int(val) for val in (min(x0, x1), min(y0, y1), w, h)]
+
+        self.ctx.new_path()
+        self.ctx.set_line_width(0.5)
+        self.ctx.rectangle(rect[0], rect[1], rect[2], rect[3])
+        self.ctx.set_source_rgb(0, 0, 0)
+        self.ctx.stroke()
+
+ToolRubberband = RubberbandGTK3
+
+
+class ToolbarGTK3(ToolbarBase, Gtk.Box):
+    def __init__(self, manager):
+        ToolbarBase.__init__(self, manager)
+        Gtk.Box.__init__(self)
+        self.set_property("orientation", Gtk.Orientation.VERTICAL)
+
+        self._toolbar = Gtk.Toolbar()
+        self._toolbar.set_style(Gtk.ToolbarStyle.ICONS)
+        self.pack_start(self._toolbar, False, False, 0)
+        self._toolbar.show_all()
+        self._toolitems = {}
+        self._signals = {}
+        self._setup_message_area()
+
+    def _setup_message_area(self):
+        box = Gtk.Box()
+        box.set_property("orientation", Gtk.Orientation.HORIZONTAL)
+        sep = Gtk.Separator()
+        sep.set_property("orientation", Gtk.Orientation.VERTICAL)
+        box.pack_start(sep, False, True, 0)
+        self.message = Gtk.Label()
+        box.pack_end(self.message, False, False, 0)
+        self.pack_end(box, False, False, 5)
+        box.show_all()
+
+        sep = Gtk.Separator()
+        sep.set_property("orientation", Gtk.Orientation.HORIZONTAL)
+        self.pack_end(sep, False, True, 0)
+        sep.show_all()
+
+    def add_toolitem(self, name, group, position, image_file, description,
+                     toggle):
+        if group is None:
+            return
+
+        if toggle:
+            tbutton = Gtk.ToggleToolButton()
+        else:
+            tbutton = Gtk.ToolButton()
+        tbutton.set_label(name)
+
+        if image_file is not None:
+            image = Gtk.Image()
+            image.set_from_file(image_file)
+            tbutton.set_icon_widget(image)
+
+        if position is None:
+            position = -1
+        self._toolbar.insert(tbutton, -1)
```

Comment:
Ups, this can not be done so easily, I have to implement the groups first, passing the "raw" position, causes the buttons to mixed all over the place.